src/extensions/agent_modules/cloud_communication/adps2_mqtt_client/src/adps2_mqtt_client_module.c

```diff
@@ -100,24 +100,38 @@ static ADPS_MQTT_CLIENT_MODULE_STATE* ModuleStateFromModuleHandle(ADUC_AGENT_MOD
 ADUC_AGENT_MODULE_HANDLE ADPS_MQTT_Client_Module_Create()
 {
     ADUC_AGENT_MODULE_HANDLE handle = NULL;
+
     ADPS_MQTT_CLIENT_MODULE_STATE* moduleState = NULL;
     ADUC_AGENT_MODULE_INTERFACE* moduleInterface = NULL;
 
-    ADUC_ALLOC(moduleState);
-    ADUC_ALLOC(moduleInterface);
+    moduleState = calloc(1, sizeof(*moduleState));
+    if (moduleState == NULL)
+    {
+        goto done;
+    }
+
+    moduleInterface = calloc(1, sizeof(*moduleInterface));
+    if (moduleInterface == NULL)
+    {
+        goto done;
+    }
 
-    moduleInterface->moduleData = moduleState;
     moduleInterface->destroy = ADPS_MQTT_Client_Module_Destroy;
     moduleInterface->getContractInfo = ADPS_MQTT_Client_Module_GetContractInfo;
     moduleInterface->doWork = ADPS_MQTT_Client_Module_DoWork;
     moduleInterface->initializeModule = ADPS_MQTT_Client_Module_Initialize;
     moduleInterface->deinitializeModule = ADPS_MQTT_Client_Module_Deinitialize;
 
-    handle = moduleInterface;
+    moduleInterface->moduleData = moduleState;
     moduleState = NULL; // transfer ownership
+
+    // successful, so only now transfer moduleInterface to the output handle.
+    handle = moduleInterface;
     moduleInterface = NULL; // transfer ownership
 
 done:
+    // Always free. If successfully transferred to output handle, then these are NOOPs;
+    // otherwise, the output handle is NULL and these get freed if they were allocated.
     free(moduleState);
```

Comment:
better would be Alloc_AdpsMqttClientModuleState(), Free_AdpsMqttClientModuleState() like we do in the rest of the ADUC code.

Free_ModuleState would free both modulestate and moduleinterface.

We also have a free naming that frees the outer pointer (not just members) as well.

Reply:
tracking this and the other comments for another follow-up PR.